Add FunctionSchema based Operator Registry

bwasti opened the pull request:
Summary: This enables creation of operators with FunctionSchema and IValue

Differential Revision: D13008791

jerryzh168 commented on caffe2/core/operator.h:
```diff
@@ -120,11 +151,18 @@ class CAFFE2_API OperatorBase : public Observable<OperatorBase> {
   // TODO(jerryzh): Remove this template
   template <typename T>
   inline T* Output(int idx, DeviceType type) {
-    static_assert(
```
yeah we are deprecating this API, but it should be fine for demo purposes for now.

bwasti commented on caffe2/core/operator.cc:
```diff
@@ -56,6 +56,14 @@ OperatorBase::OperatorBase(const OperatorDef& operator_def, Workspace* ws)
 
+OperatorBase::OperatorBase(
+    const c10::FunctionSchema& fn_schema,
+    const std::vector<c10::IValue>& inputs,
```
I think using move semantics on such small datatypes (IValue is 8 bytes) will lead to potential misuse.  https://godbolt.org/z/ufro_u no compiler warnings even with -pedantic, -Wall

I have to wrap function schema in a unique pointer now to avoid a bug in nvcc, so moves don't matter

bwasti commented on caffe2/core/operator.h:
```diff
@@ -442,6 +488,19 @@ class CAFFE2_API OperatorBase : public Observable<OperatorBase> {
 
+template <>
+inline NetDef OperatorBase::GetSingleArgument<NetDef>(
+    const string& name,
```
hmm, why would I std::move it?  If I'm not mistaken, this is a lookup with zero copies.

bwasti commented on caffe2/core/operator_test.cc:
```diff
@@ -595,4 +595,64 @@ TEST(IsTestArg, non_standard) {
 
+class TestOperatorWithFunctionSchema final : public Operator<CPUContext> {
+ public:
+  TestOperatorWithFunctionSchema(const OperatorDef& def, Workspace* ws)
```
this is the old interface, I do not plan to update it

dzhulgakov commented on caffe2/core/operator.h:
```diff
@@ -31,23 +35,50 @@ typedef ObserverBase<OperatorBase> OperatorObserver;
   explicit OperatorBase(const OperatorDef& operator_def, Workspace* ws);
+  explicit OperatorBase(
+      const c10::FunctionSchema&,
+      const std::vector<c10::IValue>&,
```
nit - where did the names of arguments go?

dzhulgakov commented on caffe2/core/operator_test.cc:
```diff
@@ -595,4 +595,64 @@ TEST(IsTestArg, non_standard) {
+  }
+
+  bool RunOnDevice() override {
+    auto out = Output(0);
```
use new api? `Output(0, {}, at::dtype<float>())`

dzhulgakov commented on caffe2/core/operator.h:
```diff
@@ -946,6 +1016,34 @@ C10_DECLARE_REGISTRY(
+
+#define REGISTER_FUNCTION_SCHEMA_OPERATOR(name, inputs, outputs, impl)        \
+  C10_REGISTER_CLASS(FunctionSchemaOperatorRegistry, name, impl)              \
+  struct FunctionSchemaStorageBase##name : public FunctionSchemaStorageBase { \
```
sigh, maybe we should add a registry variant that doesn't just construct the class and instead stores value? Bootcamp task?

dzhulgakov commented on caffe2/core/operator.h:
```diff
@@ -120,11 +151,19 @@ class CAFFE2_API OperatorBase : public Observable<OperatorBase> {
+      // When you get a Tensor here it is not fully initialized
+      return BlobGetMutableTensor(outputs_.at(idx), type);
+    }
+    auto& ival = ivalue_outputs_[idx];
```
this is not sufficient:
1) you assume the output is already tensor - what  if it's not?
2) this is not the api we'd be calling - instead majority of places call `OutputTensor(int idx, at::IntList dims, at::TensorOptions options)` below which avoids reinitialization/dtype change in-place and instead creates new tensor input

in both cases we'd need to change ivalue in-place, i.e. we need a pointer to ivalue for this case

bwasti replied:
makes sense -- updated

dzhulgakov reviewed (commented):
Can you add more tests, especially if the op uses the new API to change output size?

jerryzh168 commented on caffe2/core/blob.h:
```diff
@@ -28,28 +28,32 @@ inline Tensor* BlobSetTensor(Blob* blob, const Tensor& tensor) {
+      tensor.raw_mutable_data();
+    } else {
+      // create a new Tensor when the data_type doesn't match
+      return caffe2::empty(dims, options);
```
to keep the original behavior, we are changing this to `tensor.raw_mutable_data(options.dtype());` see Line 48 of blob.h

dzhulgakov commented on caffe2/core/operator.h:
```diff
@@ -120,11 +151,26 @@ class CAFFE2_API OperatorBase : public Observable<OperatorBase> {
   // TODO(jerryzh): Remove this template
   template <typename T>
   inline T* Output(int idx, DeviceType type) {
```
Ping?

dzhulgakov commented on caffe2/core/operator.h:
```diff
@@ -120,11 +151,26 @@ class CAFFE2_API OperatorBase : public Observable<OperatorBase> {
+      return BlobGetMutableTensor(outputs_.at(idx), type);
+    }
+    auto* ival = ivalue_outputs_[idx];
+    CAFFE_ENFORCE(
```
?

dzhulgakov commented on caffe2/core/operator.h:
```diff
@@ -120,11 +151,26 @@ class CAFFE2_API OperatorBase : public Observable<OperatorBase> {
+      auto at_tensor = at::Tensor(std::move(tensor.getIntrusivePtr()));
+      *ival = IValue(at_tensor);
+    }
+    output_tensors_[idx] = caffe2::Tensor(ival->toTensor());
```
I'd move  this wrapping inside the if still to save refcount (otherwise  you still bump one when you look up tensor above (or you can get device from  at::Tensor directly

dzhulgakov reviewed (commented):
Please fix mine and Jerry's comments. Otherwise looks great!

bwasti committed:
Add FunctionSchema based Operator Registry (#13789)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13789

This enables creation of operators with FunctionSchema and IValue

Reviewed By: smessmer

Differential Revision: D13008791

fbshipit-source-id: 2b558611be75d26e17c3c2d9f2e0158201b96c33
